### PR TITLE
clean html extractor

### PR DESCRIPTION
**Pull request by pankajastro**
closes: https://github.com/astronomer/ask-astro/issues/164

Currently, We have some duplicate code in the HTML extractor, this PR aims to remove the duplicate code and reuse it from html_utils.

### CODE REVIEW

**Comment by pankajastro** [comment]
currently, retry for HTTP call is missing here as we have in airflow/include/tasks/extract/utils/html_helpers.py. I was thing to do it in separate PR

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py [hunk 1]
```diff
@@ -21,6 +22,16 @@
 schedule_interval = "0 5 * * *" if ask_astro_env == "prod" else None
 
 
+@task
+def split_docs(urls, chunk_size=100) -> list[list]:
+    chunked_urls = split.split_list(list(urls), chunk_size=chunk_size)
+    df_data = []
+    for chunk_url in chunked_urls:
+        data = urls_to_dataframe(chunk_url)
+        df_data.append([data])
```
Comment (Lee-W) [inline airflow/dags/ingestion/ask-astro-load-airflow-docs.py:31]:
```suggestion
    df_data = [
        [urls_to_dataframe(chunk_url)]
        for chunk_url in chunked_urls
    ]
```

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py [hunk 2]
```diff
@@ -37,7 +48,7 @@ def ask_astro_load_airflow_docs():
 
     extracted_airflow_docs = task(airflow_docs.extract_airflow_docs)(docs_base_url=airflow_docs_base_url)
 
-    split_md_docs = task(split.split_html).expand(dfs=[extracted_airflow_docs])
+    # split_docs = task(split.split_list)(urls=extracted_airflow_docs, chunk_size=2)
```
Comment (Lee-W) [inline airflow/dags/ingestion/ask-astro-load-airflow-docs.py:51]: Should we remove the code here? Or is this comment needed for some reason?

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 3]
```diff
@@ -23,28 +24,72 @@ def is_valid_url(url):
...
+    try:
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an HTTPError for bad responses
+        return response.content
+    except requests.RequestException:
+        logger.info("Error fetching content for %s: %s", url, url)
+        return ""
+
+
+def exclude_url(url: str, exclude_literal: list[str]) -> bool:
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/html_utils.py:37]: I'm a bit confused when I first read this name. Should we rename it to `is_excluded_url`?

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 4]
```diff
@@ -23,28 +24,72 @@ def is_valid_url(url):
...
+
+def truncate_tokens(text: str, encoding_name: str = "gpt-3.5-turbo", max_length: int = 8192) -> str | None:
+    """
+    Truncates a text string based on the maximum number of tokens.
+
+    param string (str): The input text string to be truncated.
+    param encoding_name (str): The name of the encoding model.
+    param max_length (int): The maximum number of tokens allowed. Default is 8192.
+    """
+    import tiktoken
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/html_utils.py:70]: May I know why we import tiktoken here?

File: airflow/include/tasks/split.py [hunk 5]
```diff
@@ -103,3 +103,11 @@ def split_html(dfs: list[pd.DataFrame]) -> pd.DataFrame:
     df.reset_index(inplace=True, drop=True)
 
     return df
+
+
+def split_list(urls: list, chunk_size: 0):
```
Comment (Lee-W) [inline airflow/include/tasks/split.py:108]:
```suggestion
def split_list(urls: list[str], chunk_size: int = 0) -> list[str]:
```

File: airflow/include/tasks/split.py [hunk 6]
```diff
@@ -103,3 +103,11 @@ def split_html(dfs: list[pd.DataFrame]) -> pd.DataFrame:
...
+    new_list = []
+    for i in range(0, len(urls), chunk_size):
+        new_list.append(urls[i : i + chunk_size])
```
Comment (Lee-W) [inline airflow/include/tasks/split.py:111]:
```suggestion
    new_list = [
        urls[i : i + chunk_size]
        for i in range(0, len(urls), chunk_size)
    ]
```

File: airflow/include/tasks/split.py (hunk 5 above)
Comment (pankajastro) [reply airflow/include/tasks/split.py]: return would be list[list] just fixed it PTAL

File: airflow/include/tasks/extract/utils/html_utils.py (hunk 3 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/utils/html_utils.py]: indeed, it was confusing to me too thanks for suggestions

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py (hunk 2 above)
Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-load-airflow-docs.py]: No, reason. removed

File: airflow/include/tasks/extract/utils/html_utils.py (hunk 4 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/utils/html_utils.py:70]: just dag parsing passing time will have some effect but should be significant, let me know i'm ok to move it on top too

Comment (Lee-W) [reply airflow/include/tasks/extract/utils/html_utils.py:70]: I'm not sure how heavy this lib is, but I'm ok with either way. just want to confirm it. thanks!

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 7]
```diff
@@ -23,28 +24,72 @@ def is_valid_url(url):
...
+        logger.info("Error fetching content for %s: %s", url, url)
+        return ""
+
+
+def is_excluded_url(url: str, exclude_literal: list[str]) -> bool:
+    url_path = urlparse(url).path
+    return any(literal in url_path for literal in exclude_literal)
+
+
+def remove_tags(text_content: str) -> str | None:
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/html_utils.py:42]: `remove_tags` is a bit confusing as we're not remove all the `tags`. I have 2 suggestions in mind. 

1. make `tags` configurable and set the default value to `script` and `style`
2. rename it as `clean_tags`, `clean_non_content_tags` (none of them is good. naming is hard 😞)

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 8]
```diff
@@ -69,77 +114,49 @@ def get_all_links(url, exclude_docs=None):
     return urls
 
 
-def extract_internal_url(url, exclude_docs=None):
-    """
-    Recursively extract all valid and internal links from the given URL.
+def get_internal_links(base_url: str, exclude_literal: list[str] = None) -> set[str]:
```
Comment (Lee-W) [inline airflow/include/tasks/extract/utils/html_utils.py:117]:
```suggestion
def get_internal_links(base_url: str, exclude_literal: list[str] | None = None) -> set[str]:
```

**Review by Lee-W — COMMENTED** [review]
left a few minor suggestion, but I think we're almost good to merge it

File: airflow/include/tasks/extract/utils/html_utils.py (hunk 7 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/utils/html_utils.py]: Added tags, and changed name to clean_tags

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py [hunk 9]
```diff
@@ -21,6 +23,12 @@
 schedule_interval = "0 5 * * *" if ask_astro_env == "prod" else None
 
 
+@task
+def split_docs(urls, chunk_size=100) -> list[list[pd.Dataframe]]:
```
Comment (sunank200) [inline airflow/dags/ingestion/ask-astro-load-airflow-docs.py:27]: I feel split docs as a separate task looks better for user

File: airflow/include/tasks/extract/airflow_docs.py [hunk 10]
```diff
@@ -1,36 +1,13 @@
...
-
-    param docs_base_url: Base URL to start extract.
-    type docs_base_url: str
-
-    The returned data includes the following fields:
-    'docSource': 'apache/airflow/docs'
-    'docLink': URL for the page
-    'content': HTML content of the page
-    'sha': A UUID from the other fields
+    This task return all internal url for Airflow docs
```
Comment (sunank200) [inline airflow/include/tasks/extract/airflow_docs.py:10]: If you look at diff the previous doc string looks better. We are missing many context for user about what we are doing.

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 11]
```diff
@@ -23,28 +24,74 @@ def is_valid_url(url):
     return bool(parsed.netloc) and bool(parsed.scheme)
 
 
-def exclude_path(url, exclude_docs=None):
+def fetch_page_content(url: str) -> str:
+    try:
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:28]: missing doc string and typing

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 12]
```diff
@@ -23,28 +24,74 @@ def is_valid_url(url):
...
+        response = requests.get(url)
+        response.raise_for_status()  # Raise an HTTPError for bad responses
+        return response.content
+    except requests.RequestException:
+        logger.info("Error fetching content for %s: %s", url, url)
+        return ""
+
+
+def is_excluded_url(url: str, exclude_literal: list[str]) -> bool:
+    url_path = urlparse(url).path
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:38]: missing doc string

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 13]
```diff
@@ -23,28 +24,74 @@ def is_valid_url(url):
...
+
+
+def clean_tags(text_content: str, tags: list[str] | None) -> str | None:
     """
-    Check if the URL path contains any of the specified strings to be excluded.
+    Clean the HTML content by removing script and style tags, collapsing whitespaces, and extracting text.
 
-    param url (str): The URL to check.
-    param exclude_docs (list): List of strings to exclude from the URL path.
+    param text_content (str): The HTML content to be cleaned.
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:46]: mssing params `tags`

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 14]
```diff
@@ -23,28 +24,74 @@ def is_valid_url(url):
...
 
-def get_all_links(url, exclude_docs=None):
+
+def truncate_tokens(text: str, encoding_name: str = "gpt-3.5-turbo", max_length: int = 8192) -> str | None:
+    """
+    Truncates a text string based on the maximum number of tokens.
+
+    param string (str): The input text string to be truncated.
+    param encoding_name (str): The name of the encoding model.
+    param max_length (int): The maximum number of tokens allowed. Default is 8192.
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:70]: Why is default 8192? Any idea?

File: airflow/include/tasks/extract/utils/html_utils.py (hunk 4 above)
Comment (sunank200) [reply airflow/include/tasks/extract/utils/html_utils.py:72]: Yes lets test it @pankajastro

File: airflow/include/tasks/split.py [hunk 15]
```diff
@@ -103,3 +103,7 @@ def split_html(dfs: list[pd.DataFrame]) -> pd.DataFrame:
     df.reset_index(inplace=True, drop=True)
 
     return df
+
+
+def split_list(urls: list[str], chunk_size: int = 0) -> list[list]:
```
Comment (sunank200) [inline airflow/include/tasks/split.py:108]: missing doc strings

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 16]
```diff
@@ -69,77 +116,49 @@ def get_all_links(url, exclude_docs=None):
...
 
 
-def url_to_df(urls, doc_source=""):
+def urls_to_dataframe(
+    urls: set[str], doc_source: str = "", clean_tag: bool = True, truncate_text: bool = True
+) -> pd.DataFrame:
     """
     Create a DataFrame from a list of URLs by processing each URL and organizing the results.
 
     param urls (list): A list of URLs to be processed.
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:155]: missing params

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 17]
```diff
@@ -69,77 +116,49 @@ def get_all_links(url, exclude_docs=None):
...
-        logger.info("Error fetching content for %s: %s", url, url)
-        return None
-
-
-def process_url(url, doc_source=""):
+def process_url(url, doc_source="", clean_tag: bool = True, truncate_text: bool = True) -> dict | None:
     """
     Process a URL by fetching its content, cleaning it, and generating a unique identifier (SHA) based on the cleaned content.
 
     param url (str): The URL to be processed.
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:135]: missing params

File: airflow/include/tasks/extract/utils/html_utils.py [hunk 18]
```diff
@@ -69,77 +116,49 @@ def get_all_links(url, exclude_docs=None):
     return urls
 
 
-def extract_internal_url(url, exclude_docs=None):
-    """
-    Recursively extract all valid and internal links from the given URL.
+def get_internal_links(base_url: str, exclude_literal: list[str] | None = None) -> set[str]:
+    if exclude_literal is None:
+        exclude_literal = []
```
Comment (sunank200) [inline airflow/include/tasks/extract/utils/html_utils.py:122]: missing doc strings and params

**Review by sunank200 — CHANGES_REQUESTED** [review]
@pankajastro Can you run this again with new weaviate provider by deleting the weaviate index and reingesting, please?

The code in main branch uses latest weaviate provider. You might want to rebase and re-run the DAG after deleting the weaviate index. First you need to run bulk ingest then you need to run other DAGs.

File: airflow/include/tasks/extract/utils/html_utils.py (hunk 14 above)
Comment (pankajastro) [reply airflow/include/tasks/extract/utils/html_utils.py:71]: It depends on embedding model, In our case I have notice if token size is bigger than 8192 it fail

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py (hunk 9 above)
Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-load-airflow-docs.py]: what do you mean by "split docs as a separate"? This is a task.

Comment (sunank200) [reply airflow/dags/ingestion/ask-astro-load-airflow-docs.py]: I think this DAG just shows two tasks extract and ingest. There should also be split

**Review by sunank200 — APPROVED** [review]

File: airflow/dags/ingestion/ask-astro-load-airflow-docs.py (hunk 9 above)
Comment (pankajastro) [reply airflow/dags/ingestion/ask-astro-load-airflow-docs.py]: No, it should be 3
Screenshot 2024-01-11 at 3 34 50 PM

**Pull request merged**